scripts/testnet.sh

```diff
@@ -46,18 +49,84 @@ get_rpc() {
   echo "${unresolved_rpc}" | envsubst
 }
 
+# $1 = environment
+# $2 = token (native | hopr)
+funding_wallet_info() {
+  local environment="${1}"
+  local token="${2}"
+  curl --silent "$API_ENDPOINT/api/faucet/$environment/info?text=$token"
+}
+
+# $1 = environment
+# $2 = address
+# $3 = token
+wallet_balance() {
+  local environment="${1}"
+  local address="${2}"
+  curl --silent "$API_ENDPOINT/api/balance/$environment/$address/$token?text=true"
+}
+
+# $1 = environment
+# $2 = address
+# $3 = token (native | hopr)
+faucet_to_address() {
+  local environment="${1}"
+  local address="${2}"
+  local token="${3}"
+  local secret="${FAUCET_SECRET_API_KEY}"
+
+  curl --silent --request POST \
+  "$API_ENDPOINT/api/faucet/$environment/$address/$token?text=true" \
+  --header 'Content-Type: application/json' \
+  --data-raw "{\"secret\": \"$secret\"}"
+}
+
 # $1=account (hex)
 # $2=environment
 fund_if_empty() {
   local address="${1}"
   local environment="${2}"
 
-  # start funding in sequence (nonce-tracker will not be able to keep track of nonce in two processes)
-  # we need to use yarn explicitely to ensure packages can be resolved properly
-  local fundNativeCmd="yarn --silent run ts-node ${mydir}/fund-address.ts \
-	  --environment ${environment} --address ${address} --targetERC20 ${min_funds_hopr} --targetNative ${min_funds}"
-  # wait 15 seconds in between retries to prevent race conditions
-  PRIVATE_KEY=${FUNDING_PRIV_KEY} try_cmd "${fundNativeCmd}" 3 15
+  local faucet_address
+  faucet_address        = $(funding_wallet_info "${environment}" "address")
+
+  local faucet_native_balance faucet_hopr_balance
+  faucet_native_balance = $(funding_wallet_info "${environment}" "native")
+  faucet_hopr_balance   = $(funding_wallet_info "${environment}" "hopr")
+
+  if [ "${faucet_native_balance}" = '0.0' ]; then
+    log "Wallet ${faucet_address} has zero balance and cannot fund node ${address}"
+    exit 1
+  fi;
+
+  if [ "${faucet_hopr_balance}" = '0.0' ]; then
+    log "Wallet ${faucet_address} has no HOPR tokens and cannot fund node ${address}"
+    exit 1
+  fi;
+
+  log "Funding wallet ${faucet_address} has enough funds: ${faucet_native_balance}"
+  log "Funding wallet ${faucet_address} has HOPR tokens: ${faucet_hopr_balance}"
+
+  local address_native_balance address_hopr_balance
+  log "Checking balance of the address to fund: ${address}"
+  address_native_balance=$(wallet_balance "${environment}" "${address}" "native")
+  log "Checking balance of the address to fund: ${address}"
+  address_hopr_balance=$(wallet_balance "${environment}" "${address}" "native")
+
+  log "Native balance of ${address} is ${address_native_balance}"
+  log "HOPR balance of ${address} is ${address_hopr_balance}"
+
+  if [ "${address_native_balance}" = '0.0' ]; then
```

Comment:
Apart from the first attempt we will probably never reach the exact zero anymore. I think having the top up functionality is very useful to keep nodes funded.

Reply:
Zero only happens to new nodes. Not sure what do you mean by top-up functionality?